source/robot_model/src/Model.cpp

```diff
@@ -411,6 +412,17 @@ Model::inverse_velocity(const std::vector<state_representation::CartesianTwist>&
   dX.tail(6) = cartesian_twists.back().data();
   jacobian.bottomRows(6) = this->compute_jacobian(joint_positions, frames.back()).data();
 
+  // add damped least square term
+  if (jacobian.rows() > jacobian.cols()){
+    jacobian = jacobian.transpose().completeOrthogonalDecomposition().pseudoInverse() * 
+                (jacobian.transpose() * jacobian + 
+                dls_lambda * dls_lambda * Eigen::MatrixXd::Identity(jacobian.cols(), jacobian.cols()));
+  } else {
+    jacobian = (jacobian * jacobian.transpose() + 
+                dls_lambda * dls_lambda * Eigen::MatrixXd::Identity(jacobian.rows(), jacobian.rows())) *
+                jacobian.transpose().completeOrthogonalDecomposition().pseudoInverse();
```

Comment:
This looks incorrect to me. The first block $JJ^T + \lambda^2 I$ will be a square matrix with size `jacobian.rows()`. But then you multiply it by the Jacobian transpose, so for any non-square Jacobian you'll have an incompatible matrix product.

i.e. if the Jacobian is of size 6x7 (very common case with 7-joint robots), then this operation would be 6x6 * 7x6

This is based on your choice of `if (jacobian.rows() > jacobian.cols())` above, where this branch includes any Jacobians where `jacobian.rows() <= jacobian.cols()`

Reply:
> This looks incorrect to me. The first block will be a square matrix with size jacobian.rows(). But then you multiply it by the Jacobian transpose, so for any non-square Jacobian you'll have an incompatible matrix product.

Lastly, I looked at this realised the issue. I am not multiplying with the jacobian transpose but the jacobian transpose pseudoinverse. Here I have some scratchings of why I think the maths is fine.

![image](https://github.com/aica-technology/control-libraries/assets/77909051/2e2bf6a9-0ec1-41a3-a053-04362ae0d7c7)

Reply:
Ah ok, very good point, I forgot that the pseudo inverse also acts as a transposition. Thanks for following up!